### DISCUSSION TITLE
Image Stacking

### DISCUSSION THREAD

**Discussion by StarGeezerPhil** [aaronwmorris/indi-allsky#502] [Q&A]
Hi @aaronwmorris,

Excited to see the introduction of image stacking, but could you provide a quick overview on how it works/what it does if I enable it or how you envision it being used?

E.g. 
- does it replace the normal images saved, or is it in addition to those images
- can I stack the whole night (I've done this with DSS separately before)
- are these images just used for the star trails / star trail timelapse?

Awesome to see you adding so much - need to really catch up on developments. Do you post a summary changelog for each release iteration?

**Comment by aaronwmorris** [comment]
I need to do a much better job with the changelog.  :-)  I think I need to just create a section in the readme to highlight new features.

The stacking is performed as kind of a rolling stack.  The system keeps X number of previous images in the buffer and then stacks these images together.  If you enable registration/alignment, then the last image is used as the reference for alignment.

The stacked image does replace the original image, so it is used as the basis for the timelapse video and the startrails/keograms.

The results of stacking (and alignment) is a mixed bag.  On my monochrome camera, it increased the detected stars by at least a factor of 2.  I had star counts increase from ~400 to over 1000 in some cases.  The background of the images is significantly improved, especially with color cameras--the chrominance noise is significantly reduced in high gain situations.

Stacking is a relatively inexpensive operation, even on low end hardware.  Registration is much more expensive, assuming it even works.  Given favorable conditions (few clouds, etc) it can take a few seconds.  However, I have seen the alignment runaway with 45 seconds of time and fail to find a solution.  I have had to put in a few measures to ensure the alignment does cause the system to fall significantly behind on processing.

Stacking a whole night will probably not produce the results you want, at least not using the software packages I use for indi-allsky.  The fisheye lenses common with allsky cameras have a lot of distortion that does not stack very well for large image sets.  I have a couple of test scripts in the repo for doing this, but it takes a significant amount of time.

I posted this on cloudynights... this is 1 hour of data stacked together:
![fisheye8](https://user-images.githubusercontent.com/17464290/210176776-ecac7c61-e2de-4525-9ef0-eb305bf58dda.png)

**Comment by dbainbridge** [comment]
Is it possible to have an option NOT to replace the original?  I don't know if it is too much but would almost like two separate flows.. One with the stacked and one without.  I do not want originals replaced.  I tried stacking once already but did not like the split screen and loss of original data.

**Reply by aaronwmorris** [reply]
I am not a fan of losing the original data either.  Things like meteors are effectively "averaged out" when stacking.

I will think on it.  I may add an option to save to JPEG in with the "save raw" option.  The original image would not be processed (rotated, flipped, etc), but at least you could keep an original image.

**Comment by aaronwmorris** [comment]
Merged #504 to export non-stacked images.  JPEG is now an option for the "raw" export even though it is not technically raw.  The "RAW" image is unaltered, [dark frame] calibrated data.

**Comment by StarGeezerPhil** [comment]
Thanks @aaronwmorris,

I've barely had any clear skies to really test this out yet, but I think my results are probably:

- **Individual images are better**; though, I seem to be getting a fair amount of misalignment. Any tips
- **Timelapses aren't**; due to the stacking the resulting timelapses loose their quality - appearing jumpy/blurred as a image feature (e.g. passing cloud) lasts 3+ frames - so the smoothness is lost entirely.

Therefore, if there's anyway to have a set of high quality individual images stacked, but use individual frames for the timelapses I suspect this might give the best results.

However, as I say - only one clear(-ish) night so far since testing and the Moon is high/bright - so not the best test cases so far.

**Reply by aaronwmorris** [reply]
I still have some work to do with the alignment.  Aligning allsky images appears to be basically a worst case scenario for the stacking library.

I have been considering implementing something like the ADU (sky brightness) or star count thresholds.  Perhaps if those values go below/above a certain threshold, then stacking is halted.

I am also looking at establishing an average rotation value and eliminating images when the rotation exceeds an average threshold.  This could hopefully eliminate a lot of the misalignments.

**Reply by aaronwmorris** [reply]
You may want to try the latest code.  I made significant improvements to detecting bad rotation transforms.

**Reply by StarGeezerPhil** [reply]
Thanks @aaronwmorris - just updated and will see how it goes. 

Immediately, it looks like it's highlighting any hot/cold pixels, but will let it run and see how it goes.

**Reply by aaronwmorris** [reply]
Hot pixels will definitely interfere with the algorithm.  Darks are an absolute must (since dithering is not really an option).

**Reply by StarGeezerPhil** [reply]
Have you noticed this elsewhere @aaronwmorris... ?

When using the stack feature my images approaching/leaving dawn and dusk periods become very grey:

The rest of the time the images are absolutely fine

![image](https://user-images.githubusercontent.com/94135145/220169195-b6542c09-c3af-43e6-a34a-4b084af9193f.png)

Also visible in keograms:

![image](https://user-images.githubusercontent.com/94135145/220169382-6dac617c-0027-4c27-a5b0-fedf34fa00e2.png)

**Reply by aaronwmorris** [reply]
I have not noticed that.  If you are using averaging, it could just be the average due to the sunlight

**Reply by StarGeezerPhil** [reply]
This is using the Maximum method?

Same again yesterday:

![image](https://user-images.githubusercontent.com/94135145/220346013-6872a6e7-ce8a-4909-99b4-83960bcef36c.png)

**Reply by aaronwmorris** [reply]
Does it do that with stacking turned off?

**Reply by StarGeezerPhil** [reply]
Hi Aaron, no - it only seems to happen when I activate stacking.

However, seems hit and miss: didn't do it yesterday, or Sunday but happened all other days this last week.

I might disable stacking again until I can sort out some fresh darks anyway. I've pushed gain further than when I originally did the darks, so get a lot more noise now.

**Reply by aaronwmorris** [reply]
It is possible the gray is being caused by my image scaling algorithm.  I have been needing to take a look at that again.

**Reply by aaronwmorris** [reply]
I have merged #596 that might help with this.  Previously, I scaled images to 8 bit depth based on the detected bit depth of each image.  With this change, the maximum detected bit depth during the session is used for all images for scaling.

**Reply by StarGeezerPhil** [reply]
Hmmm... just to confirm: I've since disabled stacking again but the issue persists, e.g.

![image](https://user-images.githubusercontent.com/94135145/222786699-1a50b753-364c-4641-8907-644e49fa8ee0.png)

However, I have not updated in the last few days since your last message - so will try this for tonight/tomorrow.

**Reply by StarGeezerPhil** [reply]
New error on updating - `git pull origin master` results in reconcile warning and fatal error:

`fatal: couldn't find remote ref master`

I'll take a look in issues to see if this is already reported, but haven't seen this before.

@aaronwmorris

**Reply by aaronwmorris** [reply]
The master branch has been renamed `main`.

https://github.com/aaronwmorris/indi-allsky/discussions/501

**Reply by StarGeezerPhil** [reply]
Thanks aaron - so `git pull origin main` I guess?

**Reply by aaronwmorris** [reply]
Click on that link, there are a few commands to completely rename things in your local clone.

**Reply by StarGeezerPhil** [reply]
Hmmm.. since updating I'm getting the system going down repeatedly. Out and about today, so I'll take a proper look when I can.

But, looks like the issues may reside somewhere here:

```
Mar  4 13:41:52 SkyCam [ERROR] MainProcess allsky.unhandled_exception() #41: '  File "/home/pi/indi-allsky/virtualenv/indi-allsky/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute\n    cursor.execute(statement, parameters)\n'
Mar  4 13:41:52 SkyCam [ERROR] MainProcess allsky.unhandled_exception() #41: '  File "/home/pi/indi-allsky/virtualenv/indi-allsky/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1900, in _execute_context\n    self.dialect.do_execute(\n'
Mar  4 13:41:52 SkyCam [ERROR] MainProcess allsky.unhandled_exception() #41: '  File "/home/pi/indi-allsky/virtualenv/indi-allsky/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 210, in raise_\n    raise exception\n'
Mar  4 13:41:52 SkyCam [ERROR] MainProcess allsky.unhandled_exception() #41: '  File "/home/pi/indi-allsky/virtualenv/indi-allsky/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 2124, in _handle_dbapi_exception\n    util.raise_(\n'
Mar  4 13:41:52 SkyCam [ERROR] MainProcess allsky.unhandled_exception() #41: '  File "/home/pi/indi-allsky/virtualenv/indi-allsky/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1943, in _execute_context\n    self._handle_dbapi_exception(\n'
Mar  4 13:41:52 SkyCam [ERROR] MainProcess allsky.unhandled_exception() #41: '  File "/home/pi/indi-allsky/virtualenv/indi-allsky/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1572, in _execute_clauseelement\n    ret = self._execute_context(\n'
Mar  4 13:41:52 SkyCam [ERROR] MainProcess allsky.unhandled_exception() #41: '  File "/home/pi/indi-allsky/virtualenv/indi-allsky/lib/python3.9/site-packages/sqlalchemy/sql/elements.py", line 334, in _execute_on_connection\n    return connection._execute_clauseelement(\n'
Mar  4 13:41:52 SkyCam [ERROR] MainProcess allsky.unhandled_exception() #41: '  File "/home/pi/indi-allsky/virtualenv/indi-allsky/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1705, in _execute_20\n    return meth(self, args_10style, kwargs_10style, execution_options)\n'
Mar  4 13:41:52 SkyCam [ERROR] MainProcess allsky.unhandled_exception() #41: '  File "/home/pi/indi-allsky/virtualenv/indi-allsky/lib/python3.9/site-packages/sqlalchemy/orm/persistence.py", line 1001, in _emit_update_statements\n    c = connection._execute_20(\n'
Mar  4 13:41:52 SkyCam [ERROR] MainProcess allsky.unhandled_exception() #41: '  File "/home/pi/indi-allsky/virtualenv/indi-allsky/lib/python3.9/site-packages/sqlalchemy/orm/persistence.py", line 237, in save_obj\n    _emit_update_statements(\n'
Mar  4 13:41:52 SkyCam [ERROR] MainProcess allsky.unhandled_exception() #41: '  File "/home/pi/indi-allsky/virtualenv/indi-allsky/lib/python3.9/site-packages/sqlalchemy/orm/unitofwork.py", line 630, in execute\n    util.preloaded.orm_persistence.save_obj(\n'
Mar  4 13:41:52 SkyCam [ERROR] MainProcess allsky.unhandled_exception() #41: '  File "/home/pi/indi-allsky/virtualenv/indi-allsky/lib/python3.9/site-packages/sqlalchemy/orm/unitofwork.py", line 456, in execute\n    rec.execute(self)\n'
Mar  4 13:41:52 SkyCam [ERROR] MainProcess allsky.unhandled_exception() #41: '  File "/home/pi/indi-allsky/virtualenv/indi-allsky/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 3544, in _flush\n    flush_context.execute()\n'
Mar  4 13:41:52 SkyCam [ERROR] MainProcess allsky.unhandled_exception() #41: '  File "/home/pi/indi-allsky/virtualenv/indi-allsky/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 210, in raise_\n    raise exception\n'
Mar  4 13:41:52 SkyCam [ERROR] MainProcess allsky.unhandled_exception() #41: '  File "/home/pi/indi-allsky/virtualenv/indi-allsky/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__\n    compat.raise_(\n'
Mar  4 13:41:52 SkyCam [ERROR] MainProcess allsky.unhandled_exception() #41: '  File "/home/pi/indi-allsky/virtualenv/indi-allsky/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 3584, in _flush\n    transaction.rollback(_capture_exception=True)\n'
Mar  4 13:41:52 SkyCam [ERROR] MainProcess allsky.unhandled_exception() #41: '  File "/home/pi/indi-allsky/virtualenv/indi-allsky/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 3444, in flush\n    self._flush(objects)\n'
Mar  4 13:41:52 SkyCam [ERROR] MainProcess allsky.unhandled_exception() #41: '  File "/home/pi/indi-allsky/virtualenv/indi-allsky/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 808, in _prepare_impl\n    self.session.flush()\n'
Mar  4 13:41:52 SkyCam [ERROR] MainProcess allsky.unhandled_exception() #41: '  File "/home/pi/indi-allsky/virtualenv/indi-allsky/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 829, in commit\n    self._prepare_impl()\n'
Mar  4 13:41:52 SkyCam [ERROR] MainProcess allsky.unhandled_exception() #41: '  File "/home/pi/indi-allsky/virtualenv/indi-allsky/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 1451, in commit\n    self._transaction.commit(_to_root=self.future)\n'
Mar  4 13:41:52 SkyCam [ERROR] MainProcess allsky.unhandled_exception() #41: '  File "<string>", line 2, in commit\n'
Mar  4 13:41:52 SkyCam [ERROR] MainProcess allsky.unhandled_exception() #41: '  File "/home/pi/indi-allsky/indi_allsky/flask/miscDb.py", line 527, in setState\n    db.session.commit()\n'
Mar  4 13:41:52 SkyCam [ERROR] MainProcess allsky.unhandled_exception() #41: '  File "/home/pi/indi-allsky/indi_allsky/allsky.py", line 966, in run\n    self._miscDb.setState(\'WATCHDOG\', int(loop_start_time))\n'
Mar  4 13:41:52 SkyCam [ERROR] MainProcess allsky.unhandled_exception() #41: '  File "/home/pi/indi-allsky/allsky.py", line 119, in <module>\n    action_func(*args_list, **kwargs_dict)\n'
Mar  4 13:41:50 SkyCam [ERROR] MainProcess allsky.unhandled_exception() #36: Value: (sqlite3.OperationalError) database is locked#012[SQL: UPDATE state SET "createDate"=?, value=? WHERE state."key" = ?]#012[parameters: ('2023-03-04 13:41:45.530340', '1677937305', 'WATCHDOG')]#012(Background on this error at: https://sqlalche.me/e/14/e3q8)
Mar  4 13:41:50 SkyCam [ERROR] MainProcess allsky.unhandled_exception() #35: Type: <class 'sqlalchemy.exc.OperationalError'>
```

**Reply by aaronwmorris** [reply]
You are not the first person to report the database locking issue.  See #593.

I suspect the journald logging is overwhelming the system.  Can you tune the logging settings from https://github.com/aaronwmorris/indi-allsky/wiki/Disk-Optimizations#journald and see if that helps?

**Reply by aaronwmorris** [reply]
Merged #601 to extend the I/O timeout for sqlite.  Hopefully, that will help.

**Reply by StarGeezerPhil** [reply]
I've applied the journald changes... I think and it ran for ~10 hours before going down again.

Just to check, I usually use nano. So instead created the directory and then added the content via `sudo nano /etc/systemd/journald.conf.d/25-volatile.conf`:

```
[Journal]
Storage=volatile
Compress=yes
RateLimitIntervalSec=30s
RateLimitBurst=10000
SystemMaxUse=20M
```

I assume this would work exactly as you intended?

**Reply by aaronwmorris** [reply]
Yes, that should work.  The last step in that was restarting journald (rebooting would be fine too).

```
sudo systemctl restart systemd-journald
```

**Reply by StarGeezerPhil** [reply]
Yes, had done both but still getting it going down. It's always been stable until this recent update.

**Reply by aaronwmorris** [reply]
Merged #601 to set the sqlite PRAGMA synchronous to off.
